core/storage/merkle/polkadot_trie_db/polkadot_node.hpp

```diff
@@ -24,26 +28,73 @@ namespace kagome::storage::merkle {
       BranchWithValue = 0b11
     };
 
-    bool is_dirty;
+    // dummy nodes are used to avoid unnecessary reads from the storage
+    virtual bool isDummy() const = 0;
+
+    // just to avoid static_casts every time you need a switch on a node type
+    Type getTrieType() const {
+      return static_cast<Type>(getType());
+    }
+
     common::Buffer key_nibbles;
     common::Buffer value;
   };
 
   struct LeafNode : public PolkadotNode {
+    LeafNode() = default;
+    LeafNode(common::Buffer key_nibbles, common::Buffer value)
+        : PolkadotNode{std::move(key_nibbles), std::move(value)} {}
+
     ~LeafNode() override = default;
 
+    bool isDummy() const override {
+      return false;
+    }
     int getType() const override;
   };
 
   struct BranchNode : public PolkadotNode {
+    BranchNode() = default;
+    explicit BranchNode(common::Buffer key_nibbles,
+                        common::Buffer value = common::Buffer{})
+        : PolkadotNode{std::move(key_nibbles), std::move(value)} {}
+
     ~BranchNode() override = default;
 
+    bool isDummy() const override {
+      return false;
+    }
     int getType() const override;
 
     uint16_t childrenBitmap() const;
+    uint8_t childrenNum() const;
+
+    // Has 1..16 children.
+    // Stores their hashes to search for them in a storage and encode them more
+    // easily.
+    std::array<std::shared_ptr<PolkadotNode>, kMaxChildren> children;
```

Comment:
shouldn't it be 
```
using NodeHash = std::array<uint8_t, 32>;
std::array<NodeHash, kMaxChildren> children;
```
?
```

Reply:
No, in this case it's extremely complicated to manage. This is used to draw nodes in the main memory, and if some children are not needed there, they become dummy nodes (which store just hashes).